content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md

```diff
@@ -0,0 +1,171 @@
+`echoserver` to ensure that we are not losing requests throughout the migration.
+
+### Step 1: Create the new `4xlarge` node group.
+
+Create the new, `4xlarge` node group by defining it and running an update.
+
+<p align="center"><img src="ng-4xlarge.svg" width="650"/></p>
+
+### Step 2: Migrate NGINX to the `4xlarge` node group.
+
+With the `4xlarge` node group created, we'll migrate the NGINX service away
+from the `2xlarge` node group over to the `4xlarge` node group by changing
+its node selector scheduling terms.
+
+This change updates the NGINX Deployment spec to require the use of
+`c5.4xlarge` nodes during scheduling, and forces a rolling update over to the
+`4xlarge` node group.
+
+NGINX is able to successfully migrate across node groups because it is
+configured with HA settings, [spread-type scheduling predicates][tutorial-ha-refs],
+and can gracefully terminate within the Kubernetes [Pod lifecycle][pod-lifecycle].
+
+<p align="center"><img src="target-ng-4xlarge.svg" width="650"/></p>
+
+### Step 3: Decommission the `2xlarge` node group.
+
+Once NGINX has been completely migrated to the `4xlarge` node group, we can
+begin decommissioning the original `2xlarge` node group which is no longer in use.
+longer in use from Kubernetes and AWS.
+
+Decommissioning the node group involves:
+
+  * Draining the Kubernetes nodes.
+  * Delete the Kubernetes nodes from the APIServer.
+  * Scale down the Auto Scaling Group to `0`.
+  * Delete the node group.
+
+Set up `kubectl` by using the `kubeconfig` from the stack output.
+
+```bash
+$ pulumi stack output kubeconfig > kubeconfig.json
+$ export KUBECONFIG=$PWD/kubeconfig.json
+```
+
+To drain the `2xlarge` node group, we'll use `kubectl drain`:
+
+```bash
+for node in $(kubectl get nodes -l beta.kubernetes.io/instance-type=t3.2xlarge -o=name); do
+    kubectl drain --force --ignore-daemonsets --delete-local-data --grace-period=10 "$node";
+done
+```
+
+After draining has completed, use `kubectl delete node` to delete the nodes from the APIServer.
+
+```bash
+for node in $(kubectl get nodes -l beta.kubernetes.io/instance-type=t3.2xlarge -o=name); do
+    kubectl delete "$node";
+done
+```
+
+Scale down the node group Auto Scaling Group completely by setting
+the `desiredCapacity: 0` and running an update:
+
+<p align="center"><img src="scale-down-ng-2xlarge.svg" width="650"/></p>
+
+Once the Auto Scaling Group has scaled down, we can delete the node group from
+AWS and the Pulumi program:
+
+<p align="center"><img src="remove-ng-2xlarge.svg" width="650"/></p>
+
+## Summary
+
+In this post, we stood up an EKS cluster with a couple of node groups, and an
+`echoserver` and NGINX workload. We then created a new, updated node group and
+migrated NGINX over to it.
+
+We achieved this node group migration with zero downtime to our apps during
+load testing and decommissioning of the original node group.
+
+<img style="max-width:none; height: 350px; width: 850px;" src="eks-migration.gif"/>
```

Comment:
These can just be regular images, as they'll be sized to scale within the container. (We generally shouldn't need to specify explicit sizes on images, but where we do, we have some CSS utility classes for that &mdash; hit me up and I can point you to some examples.)

```suggestion
![](eks-migration.gif)
```